File: frontend/src/cadGeometry.ts
import type { ModelInput } from './domain'
import {
  getSketch,
  loopPoints,
  syncUnmeshedSurface,
  writeSketch,
  type SketchGeometry,
} from './geometrySketch'
import { nextPrefixedId } from './supports'

export interface HoleDimensions {
  kind: 'circle' | 'rectangle' | 'square'
  width: number
  height: number
  radius: number
}
export interface HoleShape extends HoleDimensions {
  center: number[]
}
export const defaultHoleDimensions: HoleDimensions = {
  kind: 'circle',
  width: 0.4,
  height: 0.2,
  radius: 0.15,
}
const cross = (a: number[], b: number[], c: number[]) =>
  (b[0] - a[0]) * (c[1] - a[1]) - (b[1] - a[1]) * (c[0] - a[0])
const distance = (a: number[], b: number[]) =>
  Math.hypot(a[0] - b[0], a[1] - b[1])
const onSegment = (a: number[], b: number[], c: number[]) =>
  Math.abs(cross(a, b, c)) < 1e-10 &&
  c[0] >= Math.min(a[0], b[0]) - 1e-10 &&
  c[0] <= Math.max(a[0], b[0]) + 1e-10 &&
  c[1] >= Math.min(a[1], b[1]) - 1e-10 &&
  c[1] <= Math.max(a[1], b[1]) + 1e-10
const intersects = (a: number[], b: number[], c: number[], d: number[]) =>
  (cross(a, b, c) * cross(a, b, d) < 0 &&
    cross(c, d, a) * cross(c, d, b) < 0) ||
  onSegment(a, b, c) ||
  onSegment(a, b, d) ||
  onSegment(c, d, a) ||
  onSegment(c, d, b)
const inside = (p: number[], polygon: number[][]) => {
  let result = false
  polygon.forEach((a, i) => {
    const b = polygon[(i + 1) % polygon.length]
    if (
      a[1] > p[1] !== b[1] > p[1] &&
      p[0] < ((b[0] - a[0]) * (p[1] - a[1])) / (b[1] - a[1]) + a[0]
    )
      result = !result
  })
  return result
}
export function polygonError(points: number[][]): string | null {
  if (points.length < 3) return 'An outline needs at least three vertices.'
  if (
    points.some(
      (p) => p.length < 2 || p.some((value) => !Number.isFinite(value)),
    )
  )
    return 'Coordinates must be finite numbers.'
  for (let i = 0; i < points.length; i++) {
    if (distance(points[i], points[(i + 1) % points.length]) < 1e-9)
      return 'Two adjacent vertices overlap.'
    for (let j = i + 1; j < points.length; j++) {
      if (j === i + 1 || (i === 0 && j === points.length - 1)) continue
      if (
        intersects(
          points[i],
          points[(i + 1) % points.length],
          points[j],
          points[(j + 1) % points.length],
        )
      )
        return 'Outline edges must not cross or touch each other.'
    }
  }
  const area = points.reduce(
    (sum, p, i) =>
      sum +
      p[0] * points[(i + 1) % points.length][1] -
      points[(i + 1) % points.length][0] * p[1],
    0,
  )
  return Math.abs(area) < 1e-12 ? 'The outline must enclose an area.' : null
}
export function holeDimensionError(shape: HoleDimensions): string | null {
  const values =
    shape.kind === 'circle'
      ? [shape.radius]
      : shape.kind === 'square'
        ? [shape.width]
        : [shape.width, shape.height]
  return values.some((value) => !Number.isFinite(value) || value <= 0)
    ? 'Dimensions must be finite and greater than zero.'
    : null
}
export function shapePoints(shape: HoleShape, circleSegments = 4): number[][] {
  const [x, y] = shape.center
  if (shape.kind === 'circle')
    return Array.from({ length: circleSegments }, (_, i) => [
      x + shape.radius * Math.cos((i * 2 * Math.PI) / circleSegments),
      y + shape.radius * Math.sin((i * 2 * Math.PI) / circleSegments),
    ])
  const w = shape.width / 2,
    h = (shape.kind === 'square' ? shape.width : shape.height) / 2
  return [
    [x - w, y - h],
    [x + w, y - h],
    [x + w, y + h],
    [x - w, y + h],
  ]
}
const segmentDistance = (p: number[], a: number[], b: number[]) => {
  const dx = b[0] - a[0],
    dy = b[1] - a[1]
  const t = Math.max(
    0,
    Math.min(
      1,
      ((p[0] - a[0]) * dx + (p[1] - a[1]) * dy) / (dx * dx + dy * dy),
    ),
  )
  return distance(p, [a[0] + t * dx, a[1] + t * dy])
}
export function sketchError(sketch: SketchGeometry): string | null {
  const outer = sketch.loops.find((loop) => loop.kind === 'outer')
  if (!outer) return 'Draw a closed outer outline.'
  const outline = loopPoints(sketch, outer)
  const outerError = polygonError(outline)
  if (outerError) return outerError
  const holes = sketch.loops.filter((loop) => loop.kind === 'hole')
  const polygons = holes.map((loop) =>
    loop.shape?.kind === 'circle'
      ? shapePoints(loop.shape, 128)
      : loopPoints(sketch, loop),
  )
  for (let i = 0; i < holes.length; i++) {
    const shape = holes[i].shape,
      points = polygons[i]
    const error = shape ? holeDimensionError(shape) : polygonError(points)
    if (error) return error
    if (
      points.some((p) => !inside(p, outline)) ||
      points.some((a, k) =>
        outline.some((b, j) =>
          intersects(
            a,
            points[(k + 1) % points.length],
            b,
            outline[(j + 1) % outline.length],
          ),
        ),
      ) ||
      (shape?.kind === 'circle' &&
        outline.some(
          (a, j) =>
            segmentDistance(
              shape.center,
              a,
              outline[(j + 1) % outline.length],
            ) <= shape.radius,
        ))
    )
      return 'Keep each hole fully inside the outline without touching its edges.'
    for (let j = 0; j < i; j++) {
      const other = polygons[j],
        otherShape = holes[j].shape
      if (
        shape?.kind === 'circle' &&
        otherShape?.kind === 'circle' &&
        distance(shape.center, otherShape.center) <=
          shape.radius + otherShape.radius
      )
        return 'Holes must not overlap or touch.'
      if (
        (shape?.kind === 'circle' &&
          other.some(
            (a, k) =>
              segmentDistance(shape.center, a, other[(k + 1) % other.length]) <=
              shape.radius,
          )) ||
        (otherShape?.kind === 'circle' &&
          points.some(
            (a, k) =>
              segmentDistance(
                otherShape.center,
                a,
                points[(k + 1) % points.length],
              ) <= otherShape.radius,
          ))
      )
        return 'Holes must not overlap or touch.'
      if (
        points.some((p) => inside(p, other)) ||
        other.some((p) => inside(p, points)) ||
        points.some((a, k) =>
          other.some((b, l) =>
            intersects(
              a,
              points[(k + 1) % points.length],
              b,
              other[(l + 1) % other.length],
            ),
          ),
        )
      )
        return 'Holes must not overlap or touch.'
    }
  }
  return null
}
export function replaceOutline(
  model: ModelInput,
  points: number[][],
): ModelInput {
  const error = polygonError(points)
  if (error) throw new Error(error)
  const signedArea = points.reduce(
    (sum, p, i) =>
      sum +
      p[0] * points[(i + 1) % points.length][1] -
      points[(i + 1) % points.length][0] * p[1],
    0,
  )
  const ordered = signedArea < 0 ? [...points].reverse() : points
  const vertices = ordered.map((coordinates, i) => ({
    id: `G${i + 1}`,
    coordinates:
      model.model_family === 'shell'
        ? [...coordinates.slice(0, 2), model.nodes[0]?.coordinates[2] ?? 0]
        : coordinates.slice(0, 2),
  }))
  // A replacement domain starts without the previous domain's boundary conditions.
  return syncUnmeshedSurface(
    writeSketch(
      { ...model, constraints: [], loads: [] },
      {
        vertices,
        loops: [
          { id: 'outer', kind: 'outer', vertexIds: vertices.map((v) => v.id) },
        ],
      },
    ),
  )
}
export function upsertHole(
  model: ModelInput,
  shape: HoleShape,
  holeId?: string,
): { model: ModelInput; id: string } {
  const error = holeDimensionError(shape)
  if (error || shape.center.some((value) => !Number.isFinite(value)))
    throw new Error(error ?? 'Center coordinates must be finite.')
  // Only the active shape's dimensions participate in the stored definition.
  // A blank width from a previous Rectangle proposal must not poison a Circle.
  shape =
    shape.kind === 'circle'
      ? { ...shape, width: shape.radius * 2, height: shape.radius * 2 }
      : shape.kind === 'square'
        ? { ...shape, height: shape.width, radius: shape.width / 2 }
        : { ...shape, radius: Math.min(shape.width, shape.height) / 2 }
  const sketch = getSketch(model)
  const old = sketch.loops.find((loop) => loop.id === holeId)
  const id =
    old?.id ??
    nextPrefixedId(
      'H',
      sketch.loops.map((loop) => loop.id),
    )
  const retained = sketch.vertices.filter(
    (vertex) => !old?.vertexIds.includes(vertex.id),
  )
  const usedIds = sketch.vertices.map((vertex) => vertex.id)
  const vertices = shapePoints(shape).map((coordinates, i) => {
    const vertexId = old?.vertexIds[i] ?? nextPrefixedId('G', usedIds)
    usedIds.push(vertexId)
    return {
      id: vertexId,
      coordinates:
        model.model_family === 'shell'
          ? [...coordinates, sketch.vertices[0]?.coordinates[2] ?? 0]
          : coordinates,
    }
  })
  const loop = {
    id,
    kind: 'hole' as const,
    shape,
    vertexIds: vertices.map((vertex) => vertex.id),
  }
  const next = {
    vertices: [...retained, ...vertices],
    loops: old
      ? sketch.loops.map((item) => (item.id === id ? loop : item))
      : [...sketch.loops, loop],
  }
  const validation = sketchError(next)
  if (validation) throw new Error(validation)
  return { model: writeSketch(model, next), id }
}

File: src/nonlinear_api/cad_geometry.py
"""Validate user-drawn planar loops before they reach the native mesher."""

from __future__ import annotations

from math import cos, hypot, isfinite, pi, sin


def _cross(a, b, c):
    return (b[0] - a[0]) * (c[1] - a[1]) - (b[1] - a[1]) * (c[0] - a[0])


def _touch(a, b, p):
    return abs(_cross(a, b, p)) <= 1e-10 and all(
        min(a[i], b[i]) - 1e-10 <= p[i] <= max(a[i], b[i]) + 1e-10 for i in (0, 1)
    )


def _intersects(a, b, c, d):
    return (_cross(a, b, c) * _cross(a, b, d) < 0 and _cross(c, d, a) * _cross(c, d, b) < 0) or any(
        (_touch(a, b, c), _touch(a, b, d), _touch(c, d, a), _touch(c, d, b))
    )


def _edges(points):
    return list(zip(points, points[1:] + points[:1], strict=True))


def _inside(p, points):
    result = False
    for a, b in _edges(points):
        if (a[1] > p[1]) != (b[1] > p[1]) and p[0] < (b[0] - a[0]) * (p[1] - a[1]) / (
            b[1] - a[1]
        ) + a[0]:
            result = not result
    return result


def _segment_distance(p, a, b):
    dx, dy = b[0] - a[0], b[1] - a[1]
    t = max(0, min(1, ((p[0] - a[0]) * dx + (p[1] - a[1]) * dy) / (dx * dx + dy * dy)))
    return hypot(p[0] - a[0] - t * dx, p[1] - a[1] - t * dy)


def validate_geometry(geometry: dict) -> dict[frozenset[str], dict]:
    """Return validated circle definitions indexed by their four arc endpoints."""
    vertices, loops = geometry.get("vertices"), geometry.get("loops")
    if not isinstance(vertices, list) or not isinstance(loops, list) or not loops:
        raise ValueError("Geometry needs vertices and one closed outer loop")
    lookup = {}
    for vertex in vertices:
        if not isinstance(vertex, dict) or not isinstance(vertex.get("id"), str):
            raise ValueError("Geometry vertices need unique IDs and finite coordinates")
        coords = vertex.get("coordinates")
        if (
            not isinstance(coords, list)
            or len(coords) not in (2, 3)
            or any(not isinstance(v, (int, float)) or not isfinite(v) for v in coords)
        ):
            raise ValueError("Geometry coordinates must be finite 2D or 3D points")
        if vertex["id"] in lookup:
            raise ValueError("Geometry vertex IDs must be unique")
        lookup[vertex["id"]] = coords
    if sum(loop.get("kind") == "outer" for loop in loops if isinstance(loop, dict)) != 1:
        raise ValueError("Geometry needs exactly one outer loop")
    if len({p[2] if len(p) == 3 else 0 for p in lookup.values()}) > 1:
        raise ValueError("Surface geometry must be planar and parallel to XY")
    circles, polygons, shapes = {}, [], []
    for loop in sorted(loops, key=lambda item: item.get("kind") != "outer"):
        ids = loop.get("vertex_ids", loop.get("vertexIds"))
        if (
            not isinstance(ids, list)
            or len(ids) < 3
            or len(set(ids)) != len(ids)
            or any(key not in lookup for key in ids)
        ):
            raise ValueError("Each geometry loop needs at least three distinct existing vertices")
        points = [lookup[key][:2] for key in ids]
        shape = loop.get("shape")
        if shape:
            kind, center = shape.get("kind"), shape.get("center")
            if (
                kind not in ("circle", "rectangle", "square")
                or not isinstance(center, list)
                or len(center) != 2
                or any(not isinstance(n, (int, float)) or not isfinite(n) for n in center)
            ):
                raise ValueError("Hole shape and center are invalid")
            names = (
                ["radius"]
                if kind == "circle"
                else ["width"]
                if kind == "square"
                else ["width", "height"]
            )
            if any(
                not isinstance(shape.get(key), (int, float))
                or not isfinite(shape[key])
                or shape[key] <= 0
                for key in names
            ):
                raise ValueError("Hole dimensions must be finite and positive")
            if loop.get("kind") != "hole" or len(ids) != 4:
                raise ValueError("Parameterized holes need four defining vertices")
            x, y = center
            if kind == "circle":
                r = shape["radius"]
                expected = [[x + r * cos(i * pi / 2), y + r * sin(i * pi / 2)] for i in range(4)]
                circles[frozenset(ids)] = shape
            else:
                w, h = shape["width"] / 2, shape["width" if kind == "square" else "height"] / 2
                expected = [[x - w, y - h], [x + w, y - h], [x + w, y + h], [x - w, y + h]]
            if any(
                hypot(a[0] - b[0], a[1] - b[1]) > 1e-8 * max(1, abs(x), abs(y))
                for a, b in zip(points, expected, strict=True)
            ):
                raise ValueError("Hole dimensions do not match the stored geometry vertices")
        edges = _edges(points)
        if abs(sum(a[0] * b[1] - b[0] * a[1] for a, b in edges)) < 1e-12:
            raise ValueError("Geometry loops must enclose a nonzero area")
        for i, (a, b) in enumerate(edges):
            if hypot(a[0] - b[0], a[1] - b[1]) < 1e-9:
                raise ValueError("Adjacent geometry vertices overlap")
            for j, (c, d) in enumerate(edges):
                if j <= i + 1 or (i == 0 and j == len(edges) - 1):
                    continue
                if _intersects(a, b, c, d):
                    raise ValueError("Geometry edges must not cross or touch")
        if shape and shape["kind"] == "circle":
            x, y = shape["center"]
            points = [
                [x + shape["radius"] * cos(i * pi / 64), y + shape["radius"] * sin(i * pi / 64)]
                for i in range(128)
            ]
        polygons.append(points)
        shapes.append(shape)
    outer = polygons[0]
    for i in range(1, len(polygons)):
        points, shape = polygons[i], shapes[i]
        if not all(_inside(p, outer) for p in points) or any(
            _intersects(a, b, c, d) for a, b in _edges(points) for c, d in _edges(outer)
        ):
            raise ValueError("Holes must lie strictly inside the outer outline")
        if (
            shape
            and shape["kind"] == "circle"
            and any(
                _segment_distance(shape["center"], a, b) <= shape["radius"]
                for a, b in _edges(outer)
            )
        ):
            raise ValueError("Circle holes must not touch the outer outline")
        for j in range(1, i):
            other, other_shape = polygons[j], shapes[j]
            if (
                shape
                and other_shape
                and shape["kind"] == other_shape["kind"] == "circle"
                and hypot(
                    shape["center"][0] - other_shape["center"][0],
                    shape["center"][1] - other_shape["center"][1],
                )
                <= shape["radius"] + other_shape["radius"]
            ):
                raise ValueError("Holes must not overlap or touch")
            if (
                any(_inside(p, other) for p in points)
                or any(_inside(p, points) for p in other)
                or any(_intersects(a, b, c, d) for a, b in _edges(points) for c, d in _edges(other))
            ):
                raise ValueError("Holes must not overlap or touch")
    return circles
